### PR TITLE
[Milestone 1] Prevent users from adding questions in personal collections to dashboards in public collections

### PR DESCRIPTION
**Pull request by WiNloSt**
> **Note**
> This PR has been divided into multiple smaller PRs instead to troubleshoot [the CI failure](https://github.com/metabase/metabase/actions/runs/6656623332/job/18090276695?pr=34833) and possibly make it easier to review.
>
> Please see the new PR here https://github.com/metabase/metabase/pull/35135

<details>
<summary>
Original PR description
</summary>

This is milestone 1 for Epic: https://github.com/metabase/metabase/issues/34733

_The testing plan below is copied from the epic_

#### Testing plan for milestone 1
Dashboards
1. Add a question to a dashboard in a public collection
    **expectation**: hides all personal collections
1. Add a question to a dashboard in a personal collection
    **expectation**: shows all collections
1. move a dashboard from a personal collection to a public collection and test 1)
    **expectation**: hides all personal collections
1. move a dashboard from a public collection to a personal collection and test 2)
    **expectation**: shows all collections

Query builder

Newly created question
1. Select a dashboard for a question in a public collection
    **expectation**: shows all collections
1. Select a dashboard for a question in a personal collection
    **expectation**: hides all public collections
1. Open public collection for a question in a public collection
    **expectation**:
        - shows `+ Create a new dashboard` option
        - shows dashboards
1. Open personal collection for a question in a public collection
    **expectation**:
        - shows `+ Create a new dashboard` option
        - shows dashboards
1. Open public collection for a question in a personal collection (this only happens for the root collection, as other public collections are already hidden)
    **expectation**:
        - hides `+ Create a new dashboard` option
        - hides dashboards
1. Open personal collection for a question in a personal collection
    **expectation**:
        - shows `+ Create a new dashboard` option
        - shows dashboards
1. When the question is in a public collection and the most recently viewed dashboard is in a public collection
    **expectation**:
        - shows the most recently viewed dashboard
1. When the question is in a public collection and the most recently viewed dashboard is in a personal collection
    **expectation**:
        - shows the most recently viewed dashboard
1. When the question is in a personal collection and the most recently viewed dashboard is in a public collection
    **expectation**:
        - shows the root collection instead
1. When the question is in a personal collection and the most recently viewed dashboard is in a personal collection
    **expectation**:
        - shows the most recently viewed dashboard

Existing question
1. Move a question from a personal collection to a public collection and test 1) from the **Newly created question** section
    **expectation**: shows all collections
1. Move a question from a public collection to a personal collection and test 2) from the **Newly created question** section
    **expectation**: hides all public collections
1. Test the rest of the scenarios from the **Newly created question** section

</details>

### CODE REVIEW

File: frontend/src/metabase/containers/DashboardPicker.tsx [hunk 1]
```diff
@@ -1,7 +1,12 @@
+import type { ComponentProps } from "react";
 import type { CollectionId, DashboardId } from "metabase-types/api";
 import ItemPicker from "./ItemPicker";
 
-export interface DashboardPickerProps {
+export interface DashboardPickerProps
+  extends Pick<
+    ComponentProps<typeof ItemPicker>,
+    "showOnlyPersonalCollections" | "onOpenCollectionChange"
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/DashboardPicker.tsx:8]: This change allows TS files that use `DashboardPicker` to pass these 2 new props
1. `showOnlyPersonalCollections`
2. `onOpenCollectionChange`

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx [hunk 2]
```diff
@@ -115,21 +124,25 @@ function ItemPicker<TId>({
...
 
     return collectionItems as CollectionPickerItem<TId>[];
-  }, [openCollection, models]);
+  }, [openCollection, models, showOnlyPersonalCollections]);
 
   const crumbs = useMemo(
     () =>
       getCrumbs(openCollection, collectionsById, id => {
         setOpenCollectionId(id);
-        onOpenCollectionChange?.(id);
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:130]: This is moved to the `useEffect` above.

File: frontend/src/metabase/dashboard/components/DashboardSidebars.jsx [hunk 3]
```diff
@@ -92,12 +92,7 @@ export function DashboardSidebars({
 
   switch (sidebar.name) {
     case SIDEBAR_NAME.addQuestion:
-      return (
-        <AddCardSidebar
-          initialCollection={dashboard.collection_id}
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/DashboardSidebars.jsx:97]: Remove the dependency on this prop because I need to access the `dashboard` in `frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx` anyway.

File: frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx [hunk 4]
```diff
@@ -28,17 +30,13 @@ QuestionPicker.propTypes = {
...
 };
 
-function QuestionPicker({
-  onSelect,
-  collectionsById,
-  getCollectionIcon,
-  initialCollection,
-}) {
+function QuestionPicker({ onSelect, collectionsById, getCollectionIcon }) {
+  const dashboard = useSelector(getDashboard);
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx:36]: Not sure if it'd be better to do these instead, but I keep it simple for now.
1. `useSelector(state => getDashboard(state).collection)`
2. or `useSelector(getDashboardCollection)` (a new selector)

File: frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx [hunk 5]
```diff
@@ -51,7 +49,11 @@ function QuestionPicker({
 
   const handleSearchTextChange = e => setSearchText(e.target.value);
 
-  const collections = (collection && collection.children) || [];
+  const allCollections = (collection && collection.children) || [];
+  const isDashboardInPublicCollection = !dashboard.collection.is_personal;
+  const collections = isDashboardInPublicCollection
+    ? allCollections.filter(collection => !collection.is_personal)
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx:55]: maybe this could be easier to understand 🤔 . Any thoughts?
```suggestion
    ? allCollections.filter(collection => isInPublicCollection(collection))
```

File: frontend/src/metabase/query_builder/components/QueryModals.tsx [hunk 6]
```diff
@@ -174,7 +174,7 @@ class QueryModals extends Component<QueryModalsProps> {
       case MODAL_TYPES.ADD_TO_DASHBOARD:
         return (
           <Modal onClose={onCloseModal}>
-            <AddToDashSelectDashModal
+            <ConnectedAddToDashSelectDashModal
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/QueryModals.tsx:177]: Nothing changed, just use the new named export variable.

File: frontend/src/metabase/entities/collections/getExpandedCollectionsById.js [hunk 7]
```diff
@@ -28,7 +28,6 @@ function getExpandedCollectionsById(
           : null,
       parent: null,
       children: [],
-      is_personal: c.personal_owner_id != null,
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/collections/getExpandedCollectionsById.js:31]: Now BE is sending `is_personal` to every collection in `GET /api/collections` so, FE don't need to do this anymore.

**Review by ranquild — DISMISSED** [review]
Pass 
```
createMockCollection({
 ...,
  children: [
    createMockCollection()
  ]
})
```
in tests, we don't need this

File: frontend/src/metabase-types/api/mocks/card.ts [hunk 8]
```diff
@@ -14,25 +14,30 @@ import {
...
+    display: "table",
+    public_uuid: null,
+    dataset_query: createMockStructuredDatasetQuery(),
+    visualization_settings: createMockVisualizationSettings(),
+    result_metadata: [],
+    dataset: false,
+    can_write: true,
+    cache_ttl: null,
+    collection,
+    collection_id: collection.id as number,
```
Comment (oisincoveney) [inline frontend/src/metabase-types/api/mocks/card.ts:34]:
```suggestion
    collection_id: collection.id as CollectionId,
```

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+          collection_id: PERSONAL_SUBCOLLECTION.id as number,
+          model: "dashboard",
+        });
+
+        it("should show most recently visited dashboard if the question is in a public collection", async () => {
+          await setup({
+            mostRecentlyViewedDashboard: dashboardInPersonalSubcollection,
+          });
+
+          // breadcrumbs
```
Comment (oisincoveney) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:278]: Maybe it would make sense to have some sort of `expectBreadcrumbs` method considering you're doing the same checks throughout the tests - something like:

```
function expectBreadcrumbs({breadcrumbs}: {breadcrumbs: string[]}) {
    for (const breadcrumb in breadcrumbs) {
        expect(
            screen.getByText(breadcrumb),
          ).toBeInTheDocument();
    }
   ... or maybe even check for order as well 🙂 
}
```

File: frontend/src/metabase-types/api/mocks/card.ts (hunk 8 above)
Comment (WiNloSt) [reply frontend/src/metabase-types/api/mocks/card.ts:34]: Unfortunately, some endpoint doesn't always return `CollectionId`. This is [an example](https://github.com/metabase/metabase/pull/34827#issuecomment-1772602403), where instead of returning `collection_id: "root"`, BE just don't send `collection` or `collection_id` key at all. So the type for this property is actually `number | null`

https://github.com/metabase/metabase/blob/cef520443a69620698979cad3062020e10d182ff/frontend/src/metabase-types/api/card.ts#L17

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx (hunk 9 above)
Comment (WiNloSt) [reply frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:278]: Good suggestion. Thanks!

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+          expect(screen.getByText(DASHBOARD_AT_ROOT.name)).toBeInTheDocument();
+        });
+      });
+
+      describe("when last visited dashboard belongs to public subcollections", () => {
+        const dashboardInPublicSubcollection = createMockDashboard({
+          id: 5,
+          name: "Dashboard in public subcollection",
+          collection: SUBCOLLECTION,
+          collection_id: SUBCOLLECTION.id as number,
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:228]: commented in the previous one

File: frontend/src/metabase-types/api/mocks/card.ts (hunk 8 above)
Comment (WiNloSt) [reply frontend/src/metabase-types/api/mocks/card.ts:34]: By the way, we agreed that [we won't change BE for that](https://github.com/metabase/metabase/pull/34827#issuecomment-1776562946) (won't send `collection: <root_collection>` and `collection_id: "root"`) so the current BE behavior will be as is.

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx [hunk 11]
```diff
@@ -100,6 +102,13 @@ function ItemPicker<TId>({
   const isPickingNotCollection = models.some(model => model !== "collection");
 
   const openCollection = collectionsById[openCollectionId];
+  const isOpenCollectionInPersonalCollection = openCollection?.is_personal;
+  const hideItems =
+    showOnlyPersonalCollections && !isOpenCollectionInPersonalCollection;
+
+  useEffect(() => {
```
Comment (oisincoveney) [inline frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:109]: Maybe this is just me but I would prefer an if statement instead of the optional function call on this one

Comment (WiNloSt) [reply frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:109]: Good point. This is worth discussing with FE team.

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx [hunk 12]
```diff
@@ -115,21 +124,25 @@ function ItemPicker<TId>({
 
     const collectionItems = list
       .filter(canWriteToCollectionOrChildren)
+      .filter(
```
Comment (oisincoveney) [inline frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:127]: Does this filter statement need to be recursive, like `canWriteToCollectionOrChildren`?

File: frontend/src/metabase-types/api/mocks/card.ts (hunk 8 above)
Comment (oisincoveney) [reply frontend/src/metabase-types/api/mocks/card.ts:34]: Ah ok, understandable. Then ignore my other comment about it 😀

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx (hunk 12 above)
Comment (WiNloSt) [reply frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:127]: It doesn't need to be, the purpose of this new `filter` is to decide whether we want to render the immediate collections in the current collection. And that information could be computed without recursively traversing the whole collection tree.

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx (hunk 11 above)
Comment (WiNloSt) [reply frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:109]: I [asked the question here](https://metaboat.slack.com/archives/C505ZNNH4/p1698145752378499). Likely I'll keep the change as is. But the question already generated some useful insights. Thanks for asking!

File: frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx (hunk 5 above)
Comment (oisincoveney) [reply frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx:55]: yeah, this is a good change. It's more explicit and easier to understand for someone who hasn't been to this part of the code before

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx (hunk 9 above)
Comment (WiNloSt) [reply frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:274]: Fixed in https://github.com/metabase/metabase/pull/34833/commits/ed05becadefba6ebf6fcce6b1052bfd54d364b18

File: frontend/src/metabase/containers/AddToDashSelectDashModal.unit.spec.tsx
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal.unit.spec.tsx:1]: converted to TSX

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:1]: Any test that doesn't have a comment attached is an existing test. I added the test number to every single new test because it's hard to follow which unit test is for which test in the testing plan

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,625 @@
...
+            mostRecentlyViewedDashboard: DASHBOARD_AT_ROOT,
+          });
+
+          // breadcrumbs
+          assertBreadcrumbs([ROOT_COLLECTION]);
+          expect(screen.getByText(DASHBOARD_AT_ROOT.name)).toBeInTheDocument();
+        });
+      });
+
+      describe("when last visited dashboard belongs to public subcollections", () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:219]: test Query builder > Newly created question > #7, #9 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,625 @@
...
+        const dashboardInPublicSubcollection = createMockDashboard({
+          id: 5,
+          name: "Dashboard in public subcollection",
+          collection: SUBCOLLECTION,
+          collection_id: SUBCOLLECTION.id as number,
+          model: "dashboard",
+        });
+
+        it("should show most recently visited dashboard if the question is in a public collection", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:228]: test Query builder > Newly created question > #7 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,625 @@
...
+          // breadcrumbs
+          assertBreadcrumbs([ROOT_COLLECTION, COLLECTION, SUBCOLLECTION]);
+
+          // dashboard item
+          expect(
+            screen.getByText(dashboardInPublicSubcollection.name),
+          ).toBeInTheDocument();
+        });
+
+        it("should show the root collection if the question is in a personal collection", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:242]: test Query builder > Newly created question > #9 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+          assertBreadcrumbs([ROOT_COLLECTION]);
+
+          // Showing personal collection means we're viewing the root collection
+          expect(
+            screen.getByRole("heading", { name: PERSONAL_COLLECTION.name }),
+          ).toBeInTheDocument();
+        });
+      });
+
+      describe("when last visited dashboard belongs to personal subcollections", () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:258]: test Query builder > Newly created question > #8, #10 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,625 @@
...
+        const dashboardInPersonalSubcollection = createMockDashboard({
+          id: 5,
+          name: "Dashboard in personal subcollection",
+          collection: PERSONAL_SUBCOLLECTION,
+          collection_id: PERSONAL_SUBCOLLECTION.id as number,
+          model: "dashboard",
+        });
+
+        it("should show most recently visited dashboard if the question is in a public collection", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:267]: test Query builder > Newly created question > #8 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,625 @@
...
+            PERSONAL_SUBCOLLECTION,
+          ]);
+
+          // dashboard item
+          expect(
+            await screen.findByText(dashboardInPersonalSubcollection.name),
+          ).toBeInTheDocument();
+        });
+
+        it("should show most recently visited dashboard if the question is in a personal collection", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:285]: test Query builder > Newly created question > #10 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,625 @@
...
+        await setup();
+
+        // breadcrumbs show root collection only
+        expect(screen.getByTestId("item-picker-header")).toHaveTextContent(
+          ROOT_COLLECTION.name,
+        );
+      });
+
+      describe("question is in a public collection", () => {
+        describe('"Create a new dashboard" option', () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:334]: test Query builder > Newly created question > #3, #4 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 20]
```diff
@@ -0,0 +1,625 @@
...
+          it('should render "Create a new dashboard" option when opening the root collection (public collection)', async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:335]: test Query builder > Newly created question > #3 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,625 @@
...
+            await setup();
+
+            expect(
+              screen.getByRole("heading", {
+                name: "Create a new dashboard",
+              }),
+            ).toBeInTheDocument();
+          });
+
+          it('should render "Create a new dashboard" option when opening public subcollections', async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:345]: test Query builder > Newly created question > #3 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 22]
```diff
@@ -0,0 +1,625 @@
...
+            );
+
+            expect(
+              screen.getByRole("heading", {
+                name: "Create a new dashboard",
+              }),
+            ).toBeInTheDocument();
+          });
+
+          it('should render "Create a new dashboard" option when opening personal subcollections', async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:373]: test Query builder > Newly created question > #4 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,625 @@
...
+            expect(
+              screen.getByRole("heading", {
+                name: "Create a new dashboard",
+              }),
+            ).toBeInTheDocument();
+          });
+        });
+
+        describe("whether we should render dashboards in a collection", () => {
+          it("should render dashboards when opening the root collection (public collection)", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:403]: test Query builder > Newly created question > #3 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,625 @@
...
+            });
+
+            expect(
+              await screen.findByRole("heading", {
+                name: dashboardInRootCollection.name,
+              }),
+            ).toBeInTheDocument();
+          });
+
+          it("should render dashboards when opening public subcollections", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:422]: test Query builder > Newly created question > #3 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,625 @@
...
+            );
+
+            expect(
+              await screen.findByRole("heading", {
+                name: dashboardInPublicSubcollection.name,
+              }),
+            ).toBeInTheDocument();
+          });
+
+          it("should render dashboards when opening personal collections", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:447]: test Query builder > Newly created question > #4 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 26]
```diff
@@ -0,0 +1,625 @@
...
+            );
+
+            expect(
+              await screen.findByRole("heading", {
+                name: dashboardInPersonalCollection.name,
+              }),
+            ).toBeInTheDocument();
+          });
+
+          it("should render dashboards when opening personal subcollections", async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:472]: test Query builder > Newly created question > #4 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 27]
```diff
@@ -0,0 +1,625 @@
...
+              await screen.findByRole("heading", {
+                name: dashboardInPersonalSubcollection.name,
+              }),
+            ).toBeInTheDocument();
+          });
+        });
+      });
+
+      describe("question is in a personal collection", () => {
+        describe('"Create a new dashboard" option', () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:523]: test Query builder > Newly created question > #5, #6 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx [hunk 28]
```diff
@@ -0,0 +1,625 @@
...
+          it('should not render "Create a new dashboard" option when opening the root collection (public collection)', async () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx:524]: test Query builder > Newly created question > #5 from [the epic](https://github.com/metabase/metabase/issues/34733)

File: frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx (hunk 5 above)
Comment (WiNloSt) [reply frontend/src/metabase/dashboard/components/add-card-sidebar/QuestionPicker.jsx:55]: Fixed in df8f53ce705c0e632a1c64373d6be0cc3dfc0cbb

File: frontend/src/metabase/dashboard/reducers.js [hunk 29]
```diff
@@ -142,12 +143,13 @@ const dashboards = handleActions(
...
-        assocIn(
-          state,
-          [payload.dashboard.id, "collection_id"],
-          payload.dashboard.collection_id,
-        ),
+      next: (state, { payload }) => {
+        return produce(state, draftState => {
+          const draftDashboard = draftState[payload.dashboard.id];
+          draftDashboard.collection_id = payload.dashboard.collection_id;
+          draftDashboard.collection = payload.dashboard.collection;
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/reducers.js:150]: handle moving dashboard to another collection type (public/personal) as our components rely on `dashboard.collection` property.

File: frontend/src/metabase/entities/collections/constants.ts
Comment (WiNloSt) [inline frontend/src/metabase/entities/collections/constants.ts:1]: update the constant to reflect new props from BE.

File: frontend/src/metabase/entities/collections/getExpandedCollectionsById.js [hunk 30]
```diff
@@ -54,7 +53,6 @@ function getExpandedCollectionsById(
       id: userPersonalCollectionId,
       parent: collectionsById[ROOT_COLLECTION.id],
       children: personalCollection?.children || [],
-      is_personal: true,
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/collections/getExpandedCollectionsById.js:57]: moved to `frontend/src/metabase/entities/collections/constants.ts`

File: frontend/src/metabase/entities/collections/getExpandedCollectionsById.js [hunk 31]
```diff
@@ -63,7 +61,6 @@ function getExpandedCollectionsById(
     ...PERSONAL_COLLECTIONS,
     parent: collectionsById[ROOT_COLLECTION.id],
     children: [],
-    is_personal: true,
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/collections/getExpandedCollectionsById.js:66]: moved to `frontend/src/metabase/entities/collections/constants.ts`

**Review by cdeweyx — APPROVED** [review]
Working as expected 👍
